src/shims/foreign_items.rs

```diff
@@ -102,6 +103,39 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
     }
 
+    /// Lookup the body of a non-external foreign item.
+    fn lookup_non_external_foreign_item(
+        &self,
+        abi: Abi,
+        link_name: Symbol,
+    ) -> InterpResult<'tcx, Option<&'mir mir::Body<'tcx>>> {
+        let this = self.eval_context_ref();
+        let tcx = this.tcx.tcx;
+        let mut instance = None;
+        for cnum in iter::once(LOCAL_CRATE).chain(tcx.crates().iter().copied()) {
+            for &(symbol, _) in tcx.exported_symbols(cnum) {
+                if let ExportedSymbol::NonGeneric(def_id) = symbol {
+                    let mono = ty::Instance::mono(tcx, def_id);
+                    let attrs = tcx.codegen_fn_attrs(def_id);
+                    let symbol_name = if let Some(export_name) = attrs.export_name {
+                        export_name
+                    } else if attrs.flags.contains(CodegenFnAttrFlags::NO_MANGLE) {
+                        tcx.item_name(def_id)
+                    } else {
+                        continue;
+                    };
+                    if symbol_name == link_name && instance.replace(mono).is_some() {
+                        throw_unsup_format!("multiple definitions of `{}`", link_name);
+                    }
+                }
+            }
+        }
+        instance.map(|instance| {
+            check_abi(abi, instance.ty(tcx, ty::ParamEnv::reveal_all()).fn_sig(tcx).abi())?;
```

Comment:
Are you sure this is needed? Calling a function with a MIR body will already do an ABI (and argument count) check.

Reply:
I remember the ABI test failed if this is not present. :thinking:

Reply:
Why did it not hit this check?

https://github.com/rust-lang/rust/blob/57e28ef86fdf528d1e348312f5b2775d9de2cbd0/compiler/rustc_mir/src/interpret/terminator.rs#L250

Reply:
I don't know... I think I should try to test it again.

Reply:
It failed again:
```
error: compile-fail test compiled successfully!
status: exit status: 0
command: "/home/hyd-dev/Code/rust-lang/miri/target/x86_64-unknown-linux-gnu/release/miri" "tests/compile-fail/non_extern_foreign_call_abi_mismatch.rs" "-L" "/tmp/compiletestIrPNFy" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/tmp/compiletestIrPNFy/non_extern_foreign_call_abi_mismatch.stage-id" "-A" "unused" "--edition" "2018" "-Dwarnings" "-Dunused" "--sysroot" "/home/hyd-dev/.cache/miri/HOST" "-L" "/tmp/compiletestIrPNFy/non_extern_foreign_call_abi_mismatch.stage-id.aux"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------

------------------------------------------

test [compile-fail] compile-fail/non_extern_foreign_call_abi_mismatch.rs ... FAILED
```

Reply:
Ah... so what happens is that we actually have three ABIs here:
* The caller ABI, determined by the type in the `Call` terminator.
* The "exported" callee ABI, determined by the type of the function in the `extern` block.
* The "implemented" callee ABI, determined by the type of the function at its definition site.

This leads to additional opportunities for mismatches -- not just for the ABI, also for things like the number of arguments.

We might have to add some extra checks [here](https://github.com/rust-lang/rust/blob/57e28ef86fdf528d1e348312f5b2775d9de2cbd0/compiler/rustc_mir/src/interpret/terminator.rs#L275) to ensure that the `source` in the MIR `Body` matches what we expect. I am not entirely sure what the right set of checks is... ultimately I feel like the type given in the `extern` block ought to be irrelevant, what matters is the type used for the actual call and the type of the implementation. I think that's mostly already what we do (the argument passing is done based on the information in the MIR body) -- except for the ABI.